### PR TITLE
feature/BAK 1305/use metadata when ingesting

### PR DESCRIPTION
**Pull request by diversemix**
### Things to note:
- This does not add new data - only updates the old (this is why families are created with the PUBLISHED status
- Data still needs fixing (12 errors) - I've been testing this on a CSV where I "fixed" the data

### CODE REVIEW

File: scripts/ingest_dfc/db_utils.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+engine = create_engine(
+    config.SQLALCHEMY_DATABASE_URI,
+    pool_pre_ping=True,
+    # TODO: configure as part of scaling work
+    pool_size=10,
+    max_overflow=240,
+)
+
+
+SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine, expire_on_commit=False)
```
Comment (joel-wright) [inline scripts/ingest_dfc/db_utils.py:16]: We've only changed this for the script for now, but this code should be forming the basis of a new ingest endpoint, so I think we need to be very careful modifying session defaults around expiry.

File: scripts/ingest_dfc/dfc/processor.py [hunk 2]
```diff
@@ -41,70 +40,85 @@ def ingest_row(db: Session, row: DfcRow) -> dict:
...
+    """Gets the ingestion function for ingesting a CSV.
+
+    Returns:
+        ProcessFunc: The function used to ingest the CSV file
+    """
+
+    def process(context: IngestContext, row: DfcRow) -> None:
         """Processes the row into the db."""
-        sys.stdout.write(f"Processing row: {row.row_number}: ")
+        sys.stdout.write(f"Ingesting row: {row.row_number}: ")
```
Comment (joel-wright) [inline scripts/ingest_dfc/dfc/processor.py:98]: Just out of interest, why `sys.stdout.write()` rather than `print()`?

**Review by joel-wright — APPROVED** [review]
Looks good, just the questions about taxonomy links to organisations & that odd problem with session expire_on_commit...

**Comment by joel-wright** [comment]
I was feeling generous, so I resolved the conflicts I introduced with the ruff/pyright update.

File: scripts/ingest_dfc/db_utils.py (hunk 1 above)
Comment (diversemix) [reply scripts/ingest_dfc/db_utils.py]: Now removed

File: scripts/ingest_dfc/dfc/match.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import re
+from typing import Optional, Set
+
+REGEX_ENDS_WITH_NUMBER = re.compile(r"(\D+)(\d+)$")
```
[the remark this answers is not in the document]
Comment (diversemix) [reply scripts/ingest_dfc/dfc/match.py:4]: Yeah that was what was intended - I guess we could change it if there is going to be a use case for it in the middle of a string :shrug:

File: scripts/ingest_dfc/dfc/match.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        return value.upper() == unknown_value.upper()
+
+    match = list(filter(try_case, allowed_set))
+    if len(match) > 0:
+        return match[0]
+
+    return None
+
+
+def match_unknown_value(unknown_value: str, allowed_set: Set) -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (diversemix) [reply scripts/ingest_dfc/dfc/match.py:18]: Gee thanks :wink: :+1:

File: scripts/ingest_dfc/dfc/family.py [hunk 5]
```diff
@@ -73,16 +74,9 @@ def _create_family_links(family: Family):
...
-            .join(MetadataOrganisation, MetadataOrganisation.taxonomy_name == MetadataTaxonomy.name)
-            .filter_by(taxonomy_name="default", organisation_id=org_id)
-            .first()
-        )
-        if not taxonomy:
-            raise ValueError(f"Could not find a default taxonomy for organisation {org_id}")
-        add_metadata(db, family.import_id, taxonomy, "default", row)
+
+        taxonomy = get_organisation_taxonomy(db, org_id)
+        add_metadata(db, cast(str, family.import_id), taxonomy, "default", row)
```
[the remark this answers is not in the document]
Comment (diversemix) [reply scripts/ingest_dfc/dfc/family.py:70]: The function needs a dict and the return of `get_organisation_taxonomy` is a dict- so all should be fine

File: scripts/ingest_dfc/dfc/organisation.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+from sqlalchemy.orm import Session
+from app.db.models.app.users import Organisation
+from app.db.models.law_policy.metadata import MetadataOrganisation, MetadataTaxonomy
+from scripts.ingest_dfc.utils import get_or_create, to_dict
+
+
+def create_organisation(db: Session):
+    def add_default_metadata(org: Organisation):
+        db.add(MetadataOrganisation(taxonomy_name="default", organisation_id=org.id))
```
[the remark this answers is not in the document]
Comment (diversemix) [reply scripts/ingest_dfc/dfc/organisation.py]: Fixed

File: app/db/models/law_policy/collection.py [hunk 7]
```diff
@@ -39,4 +39,4 @@ class CollectionOrganisation(Base):
 
     # FIXME: We should only have one owning organisation for a collection, so PK should
     #        probably be just `collection_import_id`
```
Comment (joel-wright) [inline app/db/models/law_policy/collection.py:41]: this fixme can go now

**Review by joel-wright — APPROVED** [review]
lgtm

**Pull request merged**